specs/interop.md

```diff
@@ -0,0 +1,188 @@
+
+A total of two transactions are required to complete a cross chain message. The first transaction is submitted
+to the local chain which authorizes execution on a remote chain. The second transaction is submitted to a
+remote chain, where the block builder SHOULD only include it if they are certain that the first transaction was
+included in the local chain.
+
+The term "block builder" is used interchangeably with the term "sequencer" for the purposes of this document but they need not be the same
+entity in practice.
+
+## The Interop Set
+
+The interop set defines the set of chains that a remote chains allows as local chains. Another way of
+saying it is that the interop set defines the set of initiating messages that are valid for a relaying
+message to be included. A relaying message MUST have an initiating message that is included in a chain
+in the interop set.
+
+The interop set is defined by chain id. Since it is impossible to enforce uniqueness of chain ids,
+social consensus MUST be used to determine the chain that represents the canonical chain id. This
+particularly impacts the block builder as they SHOULD use the chain id to assist in validation
+of relaying messages. 
+
+### Chain ID
+
+The concept of a chain id was introduced in [EIP-155](https://eips.ethereum.org/EIPS/eip-155) to prevent
+replay attacks between chains. This EIP does not specify the max size of a chain id, although
+[EIP-2294](https://eips.ethereum.org/EIPS/eip-2294) attempts to add a maximum size. Since this EIP is
+stagnant, all representations of chain ids MUST be the `uint256` type.
+
+## Interop Network Upgrade
+
+The interop network upgrade timestamp defines the timestamp at which all functionality in this document is considered
+the consensus rules for an OP Stack based network. On the interop network upgrade block, a set of deposit transaction
+based upgrade transactions are deterministically generated by the derivation pipeline in the following order:
+
+- L1 Attributes Transaction calling `setL1BlockValuesEcotone`
+- User deposits from L1
+- Network Upgrade Transactions
+  - L1Block deployment
+  - Update L1Block Proxy ERC-1967 Implementation Slot
+
+The execution payload MUST set `noTxPool` to `false` for this block.
+
+The exact definitions for these upgrade transactions are still to be defined.
+
+## Derivation Pipeline
+
+The derivation pipeline enforces invariants on safe blocks that include relaying transactions.
+
+- The relaying message MUST have a corresponding initiating message
+- The initiating message that corresponds to a relaying message MUST come from an allowed chain
+
+Blocks that contain transactions that relay cross domain messages to the remote chain where the
+initiating transaction does not exist MUST be considered invalid and MUST not be allowed by the
+derivation pipeline to be considered safe.
+
+There is no concept of replay protection within the protocol since there is no replay protection
```

Comment:
is there evidence of this
> no replay protection mechanism that fits well for all applications

app-level msg hash based replay protection seems sufficient. I can think of any significant use case that would entail multiple replay messages for a single initiated message

Reply:
Yes it simplifies the implementation (by not much - 1 mapping). But allows for strange outcomes

Reply:
> is there evidence of this

I can think of a few examples off of the top of my head:
- messages must go exactly in order, only once
- messages must be monotonic but can jump ahead by more than 1, only once
- messages can go in any order, only once
- messages can go in any order, as many times as possible

>- messages must be monotonic but can jump ahead by more than 1, only once

This is particularly useful for invalidating signed offchain messages, similar to the way permit2 does nonce invalidation

No replay protection mechanism exists that fits well for all applications similarly to how no cache invalidation mechanism works well for all applications. Application specific solutions should be adopted based on the application

What strage outcomes do you mean? Anybody that wants replay protection can use the messenger

Reply:
After reviewing the latest changes, I see that we're choosing to to separate away from a universal CDM base contract as designed between L1<>L2.

With a different messenger, diverging in behavior is fine. My main concern was around getting to an end-state where the CDM interface remains universal but it looks like we are explicitly choosing to handle L2<>L2 separately.

Reply:
This is being done explicitly to handle ether